styx-common/src/main/java/com/spotify/styx/util/GoogleIdTokenValidator.java

```diff
@@ -0,0 +1,218 @@
+    final GoogleIdToken googleIdToken;
+    try {
+      googleIdToken = verifyIdToken(token);
+    } catch (IOException e) {
+      logger.warn("failed to verify token");
+      return null;
+    }
+
+    if (googleIdToken == null) {
+      return null;
+    }
+
+    final String email = googleIdToken.getPayload().getEmail();
+
+    final Matcher matcher = EMAIL_PATTERN.matcher(email);
+    if (matcher.matches()) {
+      final String domain = matcher.group(1);
+      if (domainWhitelist.contains(domain)) {
+        logger.debug("domain {} in whitelist", domain);
+        return googleIdToken;
+      }
+    } else {
+      logger.warn("invalid email address {}", email);
+      return null;
+    }
+
+    if (validatedEmailCache.getIfPresent(email) != null) {
+      return googleIdToken;
+    }
+
+    final String projectId;
+    try {
+      projectId = checkProject(email);
+    } catch (IOException e) {
+      logger.info("cannot validate {}", email);
+      return null;
+    }
+
+    if (projectId != null) {
+      validatedEmailCache.put(email, projectId);
+      return googleIdToken;
+    }
+
+    return null;
+  }
+
+  @VisibleForTesting
+  void clearProjectCache() {
+    projectCache.clear();
+  }
+
+  private GoogleIdToken verifyIdToken(String token) throws IOException {
+    try {
+      return googleIdTokenVerifier.verify(token);
+    } catch (GeneralSecurityException e) {
+      return null;
+    }
+  }
+
+  private String checkProject(String email) throws IOException {
+    String projectId = null;
+
+    final Matcher matcher = SERVICE_ACCOUNT_PATTERN.matcher(email);
+    if (matcher.matches()) {
+      projectId = matcher.group(1);
+    }
+
+    if (projectId == null) {
+      // no projectId, could be GCE default
+      logger.debug("{} doesn't contain project id, try getting its project", email);
+      projectId = getProjectIdOfServiceAccount(email);
+    }
+
+    if (projectId == null) {
+      return null;
+    }
+
+    if (projectCache.contains(projectId)) {
+      logger.debug("hit cache for project id {}", projectId);
+      return projectId;
+    } else if (checkProjectId(projectId)) {
+      projectCache.add(projectId);
+      return projectId;
+    }
+
+    return null;
+  }
+
+  private String getProjectIdOfServiceAccount(String email) throws IOException {
+    try {
+      final ServiceAccount serviceAccount =
+          iam.projects().serviceAccounts().get("projects/-/serviceAccounts/" + email).execute();
+      return serviceAccount.getProjectId();
+    } catch (GoogleJsonResponseException e) {
+      if (e.getStatusCode() == 404) {
+        logger.debug("service account {} doesn't exist", email, e);
+        return null;
+      }
+
+      logger.info("cannot get project id for service account {}", email, e);
+      return null;
+    }
+  }
+
+  private boolean checkProjectId(String projectId) throws IOException {
+    try {
+      cloudResourceManager.projects().get(projectId).execute();
+      return true;
```

Comment:
Verify that the project parent (organization) is correct.

Reply:
Yeah, right!

Reply:
Current suggestion per offline discussion, have a white list with project/folder/organization entries and check the project ancestry against that whitelist: If any entity in the project ancestry is in the whitelist, grant access, otherwise deny.